### PR TITLE
chore: remove an additional "prepare" usage

### PR DESCRIPTION
**Pull request by eladb**
Follow up on #9410 and remove a few additional usages of `prepare`:

- In the API Gateway library we leveraged `prepare()` to record dependencies between all Deployment resources and all Method resources in the APIGW. The solution is to perform two-sided bookkeeping while methods/deployments are defined and record their dependencies in-band. We also refactored the way the code in `LatestDeployment` to be slightly more readable.
- In the Pipelines library prepare was replaced with an aspect (which is technically the drop-in alternative to `prepare()` in v2.0, for lack of a better solution at the moment).
- In the IAM library, the `Policy` resource needs to be conditionally created only if the document contains statements. To address that, we added a new protected API to `CfnResource` which is called `shouldSynthesize()`. By default it returns `true` but if it returns `false` (in a subclass), the resource will not be rendered into the cloudformation template.

Related: https://github.com/aws/aws-cdk-rfcs/issues/192

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iam/lib/policy.ts [hunk 1]
```diff
@@ -124,7 +124,18 @@ export class Policy extends Resource implements IPolicy {
         Lazy.stringValue({ produce: () => generatePolicyName(scope, resource.logicalId) }),
     });
 
-    const resource = new CfnPolicy(this, 'Resource', {
+    // this function returns `true` if the CFN resource should be included in
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-iam/lib/policy.ts:127]: What function?

File: packages/@aws-cdk/aws-iam/lib/policy.ts [hunk 2]
```diff
@@ -124,7 +124,18 @@ export class Policy extends Resource implements IPolicy {
...
+    // the cloudformation template unless `force` is `true`, if the policy
+    // document is empty, the resource will not be included.
+    const self = this;
+
+    class CfnPolicyConditional extends CfnPolicy {
+      protected shouldSynthesize() {
+        return self.force || self.referenceTaken || (!self.document.isEmpty && self.isAttached);
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-iam/lib/policy.ts:134]: Can you explain where this came from?

Comment (eladb) [reply packages/@aws-cdk/aws-iam/lib/policy.ts:136]: https://github.com/aws/aws-cdk/pull/9428/files#diff-e60351c0bfb7341b1a29307a3d444c5eL225

**Comment by eladb** [comment]
> Looks ok to me but I think @nija-at should have a look as well.

Yes, no worries. I contacted him and he'll take a look tomorrow.

File: packages/@aws-cdk/aws-apigateway/lib/deployment.ts [hunk 3]
```diff
@@ -133,31 +113,8 @@ class LatestDeploymentResource extends CfnDeployment {
     });
 
     this.api = props.restApi;
-
-    const originalLogicalId = Stack.of(this).getLogicalId(this);
-
-    this.overrideLogicalId(Lazy.stringValue({ produce: ctx => {
```
Comment (eladb) [inline packages/@aws-cdk/aws-apigateway/lib/deployment.ts:139]: This code was extracted as is into `calculateLogicalId` so it's easier to read.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 4]
```diff
@@ -627,6 +640,29 @@ export class RestApi extends RestApiBase {
    */
   public _attachMethod(method: Method) {
     this.methods.push(method);
+
+    // add this method as a dependency to all deployments defined for this api
+    // when additional deployments are added, _attachDeployment is called and
+    // this method will be added there.
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts:646]: Please copy over the comment from the original code that contains the link and quote from CF docs explaining why this dependency is needed. 
Helps posterity without having to fish for the commit description.

Comment (eladb) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts:646]: ok, no problem

**Pull request merged**